Create SmallHeap for JSStrings

justinmichaud opened the pull request:
#### 7b5f7e4443f57e7a4a1c755a49f18b50175696c1
<pre>
Create SmallHeap for JSStrings
<a href="https://bugs.webkit.org/show_bug.cgi?id=256815">https://bugs.webkit.org/show_bug.cgi?id=256815</a>
rdar://109377647

Reviewed by Yusuke Suzuki.

Create a space to allocate JSStrings from so that JSString* can fit in
32 bits (with 16-byte alignment).

This space must work even if the gigacage is disabled, or in processes like
Safari and MiniBrowser that should not allocate a full gigacage. For that reason,
we initialize this region separately.

A subsequent patch will make CompactPtr take advantage of this region,
so we lay down the foundation for that here by adding CompactPtrTypeTraits
to CompactPtr-able types.

* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
* Source/JavaScriptCore/heap/Heap.h:
(JSC::Heap::gigacageAuxiliarySpace):
* Source/JavaScriptCore/heap/IsoSubspace.cpp:
(JSC::IsoSubspace::IsoSubspace):
* Source/JavaScriptCore/heap/IsoSubspace.h:
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::loadCompactPtr):
(JSC::AssemblyHelpers::branchCompactPtr):
* Source/JavaScriptCore/llint/LLIntOfflineAsmConfig.h:
* Source/JavaScriptCore/runtime/ClassInfo.h:
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::JSString::destroy): Deleted.
(JSC::JSString::subspaceFor): Deleted.
(JSC::JSString::uninitializedValueInternal const): Deleted.
(JSC::JSString::valueInternal const): Deleted.
(JSC::JSString::JSString): Deleted.
(JSC::JSString::finishCreation): Deleted.
(JSC::JSString::create): Deleted.
(JSC::JSString::createHasOtherOwner): Deleted.
(JSC::JSString::toBoolean const): Deleted.
(JSC::JSString::canGetIndex): Deleted.
(JSC::JSString::offsetOfValue): Deleted.
(JSC::JSString::isRope const): Deleted.
(JSC::JSString::fiberConcurrently const): Deleted.
* Source/JavaScriptCore/tools/Integrity.cpp:
(JSC::Integrity::Analyzer::analyzeCell):
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType&gt;::emitLoadRTTFromFuncref):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::emitLoadRTTFromFuncref):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::WebAssemblyFunction::jsCallEntrypointSlow):
* Source/WTF/wtf/CompactPtr.h:
(WTF::BigHeapTypeTraits::encode):
(WTF::BigHeapTypeTraits::decode):
(WTF::CompactPtr::encode):
(WTF::CompactPtr::decode):
(WTF::CompactPtr::operator!=):
(WTF::CompactPtr::storage const):
* Source/WTF/wtf/DebugHeap.h:
* Source/WTF/wtf/FastMalloc.h:
(WTF::SmallHeapMalloc::malloc):
(WTF::SmallHeapMalloc::tryMalloc):
(WTF::SmallHeapMalloc::tryRealloc):
(WTF::SmallHeapMalloc::free):
* Source/WTF/wtf/Gigacage.cpp:
(Gigacage::tryMallocSmallHeap):
* Source/WTF/wtf/Gigacage.h:
(Gigacage::name):
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):
* Source/WTF/wtf/text/StringImpl.cpp:
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/UniquedStringImpl.h:
* Source/bmalloc/bmalloc/Gigacage.cpp:
(Gigacage::bmalloc::runwaySize):
(Gigacage::ensureSmallHeap):
(Gigacage::ensureGigacage):
* Source/bmalloc/bmalloc/Gigacage.h:
(Gigacage::name):
(Gigacage::maxSize):
(Gigacage::forEachKind):
(Gigacage::SmallHeapAllocatorInfo::baseAddress):
* Source/bmalloc/bmalloc/GigacageConfig.h:
* Source/bmalloc/bmalloc/GigacageKind.h:
* Source/bmalloc/bmalloc/HeapKind.h:
(bmalloc::isGigacage):
(bmalloc::gigacageKind):
(bmalloc::heapKind):
(bmalloc::isActiveHeapKindAfterEnsuringGigacage):
(bmalloc::mapToActiveHeapKindAfterEnsuringGigacage):
* Source/bmalloc/bmalloc/bmalloc.cpp:
(bmalloc::api::mallocOutOfLine):
* Tools/TestWebKitAPI/Tests/WTF/AlignedRefLogger.h:
* Tools/TestWebKitAPI/Tests/WTF/CompactRefPtr.cpp:

Canonical link: <a href="https://commits.webkit.org/264588@main">https://commits.webkit.org/264588@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/078a01f89bbca4382dab5b2ba49c48cea93340d3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7991 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8275 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8487 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9642 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8097 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7995 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10261 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8186 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10960 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8135 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9225 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7304 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9761 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6526 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7312 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14893 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6809 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7649 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7433 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10796 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/7558 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7903 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6428 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/8158 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7217 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1864 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1929 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11425 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/8378 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7644 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2013 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -49,7 +49,7 @@
 #endif
 
-#if HAVE(36BIT_ADDRESS)
+#if HAVE_36BIT_ADDRESS
```
Don't use direct `HAVE_`. Use `HAVE()` macro

Constellation commented on Source/WTF/wtf/CompactPtr.h:
```diff
@@ -33,24 +33,66 @@
+#if PLATFORM(COCOA)
+#include <wtf/spi/cocoa/MachVMSPI.h>
+#else
+#define HAVE_36BIT_ADDRESS 0
```
This is unnecessary. `HAVE()` macro handles this well.

Constellation commented on Source/WTF/wtf/CompactPtr.h:
```diff
@@ -33,24 +33,66 @@
-public:
-#if HAVE(36BIT_ADDRESS)
+struct BigHeapTypeTraits {
+#if HAVE_36BIT_ADDRESS
```
Ditto

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.h:
```diff
@@ -1963,24 +1963,30 @@ class AssemblyHelpers : public MacroAssembler {
     void loadCompactPtr(Address address, GPRReg dest)
     {
-#if HAVE(36BIT_ADDRESS)
+#if HAVE_36BIT_ADDRESS
```
Ditto. Use `HAVE()`. Do not use it directly.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.h:
```diff
@@ -1963,24 +1963,30 @@ class AssemblyHelpers : public MacroAssembler {
     Jump branchCompactPtr(RelationalCondition cond, GPRReg left, Address right, GPRReg scratch)
     {
-#if HAVE(36BIT_ADDRESS)
+#if HAVE_36BIT_ADDRESS
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/ClassInfo.h:
```diff
@@ -27,7 +27,7 @@
 #include <wtf/PtrTag.h>
 
-#if HAVE(36BIT_ADDRESS)
+#if HAVE_36BIT_ADDRESS || ENABLE_SMALL_HEAP
```
Ditto.

Constellation commented on Source/WTF/wtf/CompactPtr.h:
```diff
@@ -33,24 +33,66 @@
+    static ALWAYS_INLINE StorageType encode(const T* ptr)
+    {
+        uintptr_t intPtr = bitwise_cast<uintptr_t>(ptr);
+#if HAVE_36BIT_ADDRESS
```
Ditto.

Constellation commented on Source/WTF/wtf/CompactPtr.h:
```diff
@@ -33,24 +33,66 @@
+
+    static ALWAYS_INLINE T* decode(StorageType ptr)
+    {
+#if HAVE_36BIT_ADDRESS
```
Ditto.

hyjorc1 commented on Source/WTF/wtf/CompactPtr.h:
```diff
@@ -33,24 +33,66 @@
     // The CompactPtr algorithm relies on being able to shift
     // a 36-bit address right by 4 in order to fit in 32-bits.
     // The only way this is an ok thing to do without information
     // loss is if the if the address is always 16 bytes aligned i.e.
```
This is my typo (embarrassing). It should be `if only if` instead of `is if the if`.

Constellation commented on Source/WTF/wtf/text/UniquedStringImpl.h:
```diff
@@ -38,6 +38,13 @@ class UniquedStringImpl : public StringImpl {
     UniquedStringImpl(CreateSymbolTag, const UChar* characters, unsigned length) : StringImpl(CreateSymbol, characters, length) { }
     UniquedStringImpl(CreateSymbolTag) : StringImpl(CreateSymbol) { }
+public:
+#if ENABLE_SMALL_HEAP
```
Use `ENABLE(SMALL_HEAP)`.

Constellation commented on Source/bmalloc/bmalloc/BPlatform.h:
```diff
@@ -391,3 +393,9 @@
+#define HAVE_36BIT_ADDRESS 1
+#else
+#define HAVE_36BIT_ADDRESS 0
+#endif
```
Is this configuration correct for macCatalyst?

Constellation commented on Source/WTF/wtf/FastMalloc.h:
```diff
@@ -271,6 +275,43 @@ struct FastFree<T[]> {
     }
 };
 
+#if ENABLE_SMALL_HEAP
```
`ENABLE(SMALL_HEAP)`.

Constellation commented on Source/WTF/wtf/FastMalloc.h:
```diff
@@ -355,6 +396,108 @@ using __thisIsHereToForceASemicolonAfterThisMacro UNUSED_TYPE_ALIAS = int
     WTF_MAKE_FAST_ALLOCATED_IMPL \
 using __thisIsHereToForceASemicolonAfterThisMacro UNUSED_TYPE_ALIAS = int
 
+#if !ENABLE_SMALL_HEAP
```
!ENABLE(SMALL_HEAP)

Constellation commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1486,3 +1485,9 @@
 #define HAVE_MACH_RANGE_CREATE 1
 #endif
+
+#if CPU(ADDRESS64) && CPU(ARM64) && PLATFORM(IOS_FAMILY)
```
Ditto.

Constellation commented on Source/bmalloc/bmalloc/Gigacage.cpp:
```diff
@@ -117,6 +157,10 @@ void ensureGigacage()
             RELEASE_BASSERT(!g_gigacageConfig.ensureGigacageHasBeenCalled);
             g_gigacageConfig.ensureGigacageHasBeenCalled = true;
 
+#if ENABLE_SMALL_HEAP
```
Ditto.

Constellation commented on Source/bmalloc/bmalloc/Gigacage.cpp:
```diff
@@ -261,8 +308,13 @@ void removePrimitiveDisableCallback(void (*function)(void*), void* argument)
     bool isEnabled = g_gigacageConfig.isEnabled;
-    for (size_t i = 0; i < NumberOfKinds; ++i)
+    for (size_t i = 0; i < NumberOfKinds; ++i) {
+#if !ENABLE_SMALL_HEAP
```
Ditto.

Constellation reviewed (approved):
Let's see whether this does not cause the regression. If it is, then let's disable it until we fix the regression.

Commit:
Create SmallHeap for JSStrings
https://bugs.webkit.org/show_bug.cgi?id=256815
rdar://109377647

Reviewed by Yusuke Suzuki.

Create a space to allocate JSStrings from so that JSString* can fit in
32 bits (with 16-byte alignment).

This space must work even if the gigacage is disabled, or in processes like
Safari and MiniBrowser that should not allocate a full gigacage. For that reason,
we initialize this region separately.

A subsequent patch will make CompactPtr take advantage of this region,
so we lay down the foundation for that here by adding CompactPtrTypeTraits
to CompactPtr-able types.

* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
* Source/JavaScriptCore/heap/Heap.h:
(JSC::Heap::gigacageAuxiliarySpace):
* Source/JavaScriptCore/heap/IsoSubspace.cpp:
(JSC::IsoSubspace::IsoSubspace):
* Source/JavaScriptCore/heap/IsoSubspace.h:
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::loadCompactPtr):
(JSC::AssemblyHelpers::branchCompactPtr):
* Source/JavaScriptCore/llint/LLIntOfflineAsmConfig.h:
* Source/JavaScriptCore/runtime/ClassInfo.h:
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::JSString::destroy): Deleted.
(JSC::JSString::subspaceFor): Deleted.
(JSC::JSString::uninitializedValueInternal const): Deleted.
(JSC::JSString::valueInternal const): Deleted.
(JSC::JSString::JSString): Deleted.
(JSC::JSString::finishCreation): Deleted.
(JSC::JSString::create): Deleted.
(JSC::JSString::createHasOtherOwner): Deleted.
(JSC::JSString::toBoolean const): Deleted.
(JSC::JSString::canGetIndex): Deleted.
(JSC::JSString::offsetOfValue): Deleted.
(JSC::JSString::isRope const): Deleted.
(JSC::JSString::fiberConcurrently const): Deleted.
* Source/JavaScriptCore/tools/Integrity.cpp:
(JSC::Integrity::Analyzer::analyzeCell):
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType>::emitLoadRTTFromFuncref):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::emitLoadRTTFromFuncref):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.cpp:
(JSC::WebAssemblyFunction::jsCallEntrypointSlow):
* Source/WTF/wtf/CompactPtr.h:
(WTF::BigHeapTypeTraits::encode):
(WTF::BigHeapTypeTraits::decode):
(WTF::CompactPtr::encode):
(WTF::CompactPtr::decode):
(WTF::CompactPtr::operator!=):
(WTF::CompactPtr::storage const):
* Source/WTF/wtf/DebugHeap.h:
* Source/WTF/wtf/FastMalloc.h:
(WTF::SmallHeapMalloc::malloc):
(WTF::SmallHeapMalloc::tryMalloc):
(WTF::SmallHeapMalloc::tryRealloc):
(WTF::SmallHeapMalloc::free):
* Source/WTF/wtf/Gigacage.cpp:
(Gigacage::tryMallocSmallHeap):
* Source/WTF/wtf/Gigacage.h:
(Gigacage::name):
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):
* Source/WTF/wtf/text/StringImpl.cpp:
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/UniquedStringImpl.h:
* Source/bmalloc/bmalloc/Gigacage.cpp:
(Gigacage::bmalloc::runwaySize):
(Gigacage::ensureSmallHeap):
(Gigacage::ensureGigacage):
* Source/bmalloc/bmalloc/Gigacage.h:
(Gigacage::name):
(Gigacage::maxSize):
(Gigacage::forEachKind):
(Gigacage::SmallHeapAllocatorInfo::baseAddress):
* Source/bmalloc/bmalloc/GigacageConfig.h:
* Source/bmalloc/bmalloc/GigacageKind.h:
* Source/bmalloc/bmalloc/HeapKind.h:
(bmalloc::isGigacage):
(bmalloc::gigacageKind):
(bmalloc::heapKind):
(bmalloc::isActiveHeapKindAfterEnsuringGigacage):
(bmalloc::mapToActiveHeapKindAfterEnsuringGigacage):
* Source/bmalloc/bmalloc/bmalloc.cpp:
(bmalloc::api::mallocOutOfLine):
* Tools/TestWebKitAPI/Tests/WTF/AlignedRefLogger.h:
* Tools/TestWebKitAPI/Tests/WTF/CompactRefPtr.cpp:

Canonical link: https://commits.webkit.org/264588@main
